### ISSUE TITLE
sql: Investigate immutability of parsed expression trees

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#6469]
@RaduBerinde @petermattis 

We currently put a lot of effort into avoiding any in-place modifications to parsed `Expr` trees. For instance, the `parser.Walk` infrastructure will propagate up a  `changed` flag to copy nodes who's children have been modified. The reason for this is that we need to work on a fresh expression tree in the case where a query is retried. This approach works well, but has caused a few issues (see https://github.com/cockroachdb/cockroach/issues/6403). It also does not guarantee that the initial expression tree acts as copy-on-write, but instead copies the current tree on every write, meaning that the approach does not place an upper bound on the number of copies a single tree could be forced to perform (eg. a tree could be copied completely multiple times if it is run through multiple walk phases which each touch all leaf nodes).

We should investigate alternative approaches to this immutability, as permitting changes in-place could allow for a large reduction in allocations across expression simplification and analysis. Potential approaches might be to:
- preserve an immutable copy to clone and work off of on each retry
- preserve an immutable copy and only treat this as immutable, copying segments of this tree on changes into a new mutable tree (copy-on-write at the node level). This is strictly better than the first option, but also more complicated to get right in practice
- re-parse the query itself on each retry and treat the parsed expression tree as fully mutable

The performance implications of these alternatives are not clear, and depend heavily on the presence of retries.

**Comment by RaduBerinde** [comment]
@andreimatei 

There are also downsides to moving away from the immutability model. For one, different expressions cannot reliably share parts of trees anymore (e.g. see `splitFilter`). Avoiding that could actually lead to more copying.

> It also does not guarantee that the initial expression tree acts as copy-on-write, but instead copies the current tree on every write

It does do copy-on-write - it makes copies of the parents and ancestors of nodes that are modified.

**Comment by nvanbenschoten** [comment]
What I mean by copy-on-write (and I think is its true definition) is that once a copy is performed due to a desired mutation of an immutable structure, the result can then be considered mutable so that future modifications do not need to copy again. I don't believe this is true with our current approach.

For instance, if I have a tree `ParenExpr{ParenExpr{NumExpr}}` and I change `NumExpr` in two subsequent walk phases, we currently will make two complete copies of the tree. Ideally we should be able to know that once all (or part) of the tree has been copied from the original immutable version, the tree (or that part of the tree) no longer needs to be copied on modifications.

**Comment by RaduBerinde** [comment]
Ah - thanks for clarifying. Yes, that makes sense. We would need to maintain a "shared" flag per node.

**Comment by nvanbenschoten** [comment]
I threw together a quick proof of concept for the immutability flag on each node (which depends on the TODO(@knz) [here](https://github.com/cockroachdb/cockroach/blob/a7f7e5f26661e6c77d8ff3f379a52a8cf133a13e/sql/update.go#L201) to be implemented properly), and the results were pretty promising:

```
name                    old time/op    new time/op    delta
Select1_Cockroach-4       52.1µs ± 4%    52.6µs ± 2%     ~           (p=0.247 n=10+10)
Select2_Cockroach-4        781µs ± 2%     769µs ± 2%   -1.52%         (p=0.004 n=10+8)
Insert1_Cockroach-4        443µs ± 4%     469µs ± 7%   +5.74%         (p=0.006 n=9+10)
Insert10_Cockroach-4       709µs ± 8%     727µs ± 7%     ~           (p=0.165 n=10+10)
Insert100_Cockroach-4     2.48ms ± 5%    2.40ms ± 6%   -2.96%         (p=0.022 n=10+9)
Insert1000_Cockroach-4    18.2ms ± 2%    18.3ms ± 1%     ~             (p=0.863 n=9+9)
Update1_Cockroach-4        794µs ±10%     777µs ±10%     ~           (p=0.190 n=10+10)
Update10_Cockroach-4      1.45ms ± 9%    1.36ms ± 7%   -5.93%        (p=0.015 n=10+10)
Update100_Cockroach-4     6.13ms ± 6%    5.84ms ± 6%   -4.71%         (p=0.008 n=9+10)
Update1000_Cockroach-4    45.9ms ± 6%    45.8ms ± 5%     ~           (p=1.000 n=10+10)
Delete1_Cockroach-4        911µs ± 5%     855µs ± 3%   -6.20%          (p=0.000 n=9+9)
Delete10_Cockroach-4      1.58ms ± 3%    1.52ms ± 4%   -3.78%        (p=0.000 n=10+10)
Delete100_Cockroach-4     7.92ms ± 7%    7.76ms ± 3%     ~            (p=0.315 n=10+8)
Delete1000_Cockroach-4    80.4ms ± 7%    80.6ms ± 6%     ~           (p=0.971 n=10+10)

name                    old alloc/op   new alloc/op   delta
Select1_Cockroach-4       2.75kB ± 0%    2.76kB ± 0%   +0.59%        (p=0.000 n=10+10)
Select2_Cockroach-4       96.5kB ± 0%    95.7kB ± 0%   -0.84%          (p=0.000 n=9+9)
Insert1_Cockroach-4       23.5kB ± 0%    23.5kB ± 0%   +0.19%         (p=0.000 n=9+10)
Insert10_Cockroach-4      73.5kB ± 0%    73.6kB ± 0%   +0.03%          (p=0.042 n=9+9)
Insert100_Cockroach-4      505kB ± 0%     504kB ± 0%     ~           (p=0.739 n=10+10)
Insert1000_Cockroach-4    4.47MB ± 0%    4.47MB ± 0%     ~           (p=0.165 n=10+10)
Update1_Cockroach-4       47.4kB ± 0%    46.4kB ± 0%   -2.11%         (p=0.000 n=10+9)
Update10_Cockroach-4       130kB ± 0%     127kB ± 0%   -2.27%        (p=0.000 n=10+10)
Update100_Cockroach-4      870kB ± 0%     846kB ± 0%   -2.84%         (p=0.000 n=9+10)
Update1000_Cockroach-4    7.36MB ± 0%    7.14MB ± 0%   -2.97%        (p=0.000 n=10+10)
Delete1_Cockroach-4       44.4kB ± 0%    43.4kB ± 0%   -2.15%        (p=0.000 n=10+10)
Delete10_Cockroach-4      72.0kB ± 0%    69.1kB ± 0%   -4.10%         (p=0.000 n=10+9)
Delete100_Cockroach-4      337kB ± 0%     313kB ± 0%   -6.99%         (p=0.000 n=9+10)
Delete1000_Cockroach-4    3.53MB ± 0%    3.31MB ± 0%   -6.22%         (p=0.000 n=10+8)

name                    old allocs/op  new allocs/op  delta
Select1_Cockroach-4         50.0 ± 0%      50.0 ± 0%     ~     (all samples are equal)
Select2_Cockroach-4        1.41k ± 0%     1.39k ± 0%   -1.42%          (p=0.000 n=9+9)
Insert1_Cockroach-4          275 ± 1%       274 ± 0%     ~           (p=0.720 n=10+10)
Insert10_Cockroach-4         517 ± 0%       516 ± 0%     ~           (p=0.656 n=10+10)
Insert100_Cockroach-4      2.71k ± 0%     2.71k ± 0%     ~           (p=0.314 n=10+10)
Insert1000_Cockroach-4     24.4k ± 0%     24.4k ± 0%     ~           (p=0.126 n=10+10)
Update1_Cockroach-4          656 ± 0%       626 ± 0%   -4.57%        (p=0.000 n=10+10)
Update10_Cockroach-4       1.07k ± 0%     0.99k ± 0%   -7.03%        (p=0.000 n=10+10)
Update100_Cockroach-4      4.78k ± 0%     4.24k ± 0%  -11.18%         (p=0.000 n=8+10)
Update1000_Cockroach-4     39.3k ± 0%     34.6k ± 0%  -12.15%        (p=0.000 n=10+10)
Delete1_Cockroach-4          585 ± 0%       555 ± 0%   -5.13%        (p=0.000 n=10+10)
Delete10_Cockroach-4         840 ± 0%       764 ± 0%   -8.96%         (p=0.000 n=10+9)
Delete100_Cockroach-4      3.21k ± 0%     2.68k ± 0%  -16.38%        (p=0.000 n=10+10)
Delete1000_Cockroach-4     26.7k ± 0%     21.7k ± 0%  -18.78%         (p=0.000 n=10+8)
```

It seems like this might be the approach to take, but I still don't think it fully rules out the need for a better Statement/Expression tree equality approach.

**Comment by nvanbenschoten** [comment]
Record keeping: This change will be possible once https://github.com/cockroachdb/cockroach/pull/6735 is merged.

**Comment by nvanbenschoten** [comment]
Ever since #6735 was merged and I rebased on master, I haven't seen the same consistent performance gain from this change. I think the reason for this is that #6735 removed some of the duplicate expression tree mutation traversals, which I'd expect to be the primary causes of duplicate expression node modification (which in turn was causing multiple expression copies).

That means that a change like this is no longer an immediate boost to performance unless other mutable traversals are added. As such, I don't think it makes sense to complicate the walk infrastructure with this change until/if a perf boost can again be instantly realized. However, it also means that https://github.com/cockroachdb/cockroach/pull/6735 probably had performance implications that we never measured. I'm running benchmarks across the change now to confirm this theory.

**Comment by jordanlewis** [comment]
I'm closing this issue - it's an investigation from a while ago about performance specifically. We know AST nodes should be immutable at this point, and it's a matter of doing that work which will happen over time - a new issue can be created for that work as it becomes necessary.
